modules/cugraph/src/hypergraph.ts

```diff
@@ -0,0 +1,399 @@
+ *
+ * The transform reveals relationships between the rows and unique values. This transform is useful
+ * for lists of events, samples, relationships, and other structured high-dimensional data.  The
+ * transform creates a node for every row, and turns a row's column entries into node attributes.
+ * Every unique value within a column is also turned into a node.
+ *
+ * Edges are added to connect a row's nodes to one another. Nodes are given the attribute specified
+ * by ``nodeType`` that corresponds to the originating column name, or if a row ``eventId``.
+ *
+ * Consider a list of events. Each row represents a distinct event, and each column some metadata
+ * about an event. If multiple events have common metadata, they will be transitively connected
+ * through those metadata values. Conversely, if an event has unique metadata, the unique metadata
+ * will turn into nodes that only have connections to the event node.  For best results, set
+ * ``eventId`` to a row's unique ID, ``skip`` to all non-categorical columns (or ``columns`` to all
+ * categorical columns), and ``categories`` to group columns with the same kinds of values.
+ */
+export function hypergraphDirect<T extends TypeMap = any>(values: DataFrame<T>, {
+  columns       = null,
+  categories    = {},
+  dropNulls     = true,
+  edgeShape     = {},
+  dropEdgeAttrs = false,
+  skip          = [],
+  delim         = '::',
+  source        = 'src',
+  target        = 'dst',
+  nodeId        = 'node_id',
+  eventId       = 'event_id',
+  category      = 'category',
+  edgeType      = 'edge_type',
+  nodeType      = 'node_type',
+}: HypergraphDirectProps = {}): HypergraphReturn {
+  const computed_columns = _compute_columns(values, columns, skip);
+
+  const initial_events =
+    _create_events(values, computed_columns, delim, dropNulls, eventId, nodeType);
+
+  const entities = _create_entity_nodes(
+    initial_events, computed_columns, dropNulls, categories, delim, nodeId, category, nodeType);
+
+  const edges  = _create_direct_edges(initial_events,
+                                     computed_columns,
+                                     dropNulls,
+                                     categories,
+                                     edgeShape,
+                                     dropEdgeAttrs,
+                                     delim,
+                                     source,
+                                     target,
+                                     eventId,
+                                     category,
+                                     edgeType,
+                                     nodeType);
+  const events = new DataFrame({});
+  const nodes  = entities;
+
+  const graph = create_graph(edges, source, target);
+
+  return {nodes, edges, events, entities, graph};
+}
+
+function _compute_columns(values: DataFrame, columns: string[]|null, skip: string[]) {
```

Comment:
Could columns just be `string[]` with the default being an empty array? Likewise in the interface defined above. Similar to `skip?`

Reply:
Just from personal experience in other contexts of having done exactly that --- used "empty list is a special value that actually means grab everything available" --- I've always regretted it, and have always ended up reverting it for the sake of consistency so that the empty list just means the empty list.